draco/asp/soft.lp

```diff
@@ -30,18 +30,67 @@ preference(same_field,F) :-
     entity(mark,_,M),
     { entity(encoding,M,E): attribute((encoding,field),E,F) } = 2.
 
-% @soft(type_linear) linear scale
-preference(type_linear,E) :-
+% @soft(same_field_grt3) Prefer not to use the same field three or more times for the same mark.
+preference(same_field_grt3,F) :-
+    entity(field,_,F),
+    entity(mark,_,M),
+    { entity(encoding,M,E): attribute((encoding,field),E,F) } >= 3.
+
+% @soft(count_grt1) Prefer not to use count more than once.
+preference(count_grt1,M) :-
+    entity(mark,_,M),
+    { entity(encoding,M,E): attribute((encoding,aggregate),E,count) } > 1.
+
+% @soft(shape_high_cardinality) Shape channel should not have too high cardinality.
+preference(shape_high_cardinality,E) :-
+    attribute((encoding,channel),E,shape),
+    attribute(encoding_cardinality,E,N),
+    N > 5.
+
+% @soft(number_categorical) Numbers should not be nominal.
+preference(number_categorical,F) :-
+    attribute((field,type),F,number),
+    attribute(field_scale,F,categorical).
+
+% @soft(bin_number_low_cardinality) Binned number field should not have too low cardinality.
+preference(bin_number_low_cardinality,E) :-
+    attribute((field,type),F,number),
+    attribute((encoding,field),E,F),
+    attribute((encoding,binning),E,B),
+    B < 15.
+
+% @soft(bin_not_linear) Prefer linear scale for bin.
+preference(bin_not_linear,E) :-
+    attribute((encoding,binning),E,_),
+    not attribute((encoding,scale_type),E,linear).
+
+% @soft(only_discrete) Plots with only nominal, ordinal, binned q, or t with time unit should add either an aggregation (e.g. count) or a quantitative field.
+preference(only_discrete,M) :-
+    attribute(mark_channel_discrete_or_binned,M,_).
+
+% @soft(multi_non_pos) Prefer not to use multiple non-positional encoding channels.
+preference(multi_non_pos,M) :-
```

Comment:
I'm not sure but I think we need another constraint about non-positional channels in views since this one is about marks.

Reply:
I think using multi non-pos channels in one mark already has a penalty, then a view with multiple marks having this situation will have multiple penalties?

Reply:
I mean for example you can have one non-positional for layer 1 and a different non-positional for layer 2, which doesn't violate this particular preference. but for such view, it has 2 non-positional in total.

Reply:
I am not sure whether two non-pos in two separate layers should be penalized or not...